File: tutu-backend/src/main/scala/fi/oph/tutu/backend/domain/Hakemus.scala
package fi.oph.tutu.backend.domain

import io.swagger.v3.oas.annotations.media.Schema
import io.swagger.v3.oas.annotations.media.Schema.RequiredMode

import java.time.LocalDateTime
import java.util.UUID
import scala.annotation.meta.field

@Schema(name = "Hakemus")
case class DbHakemus(
  @(Schema @field)(
    example = "de4ffbea-1763-4a43-a24d-50ee48b81ff1",
    requiredMode = RequiredMode.REQUIRED,
    maxLength = 36
  )
  id: UUID,
  @(Schema @field)(
    example = "1.2.246.562.11.00000000000000006666",
    requiredMode = RequiredMode.REQUIRED,
    maxLength = 40
  )
  hakemusOid: HakemusOid,
  @(Schema @field)(
    example = "1",
    requiredMode = RequiredMode.REQUIRED,
    minimum = "-2147483647",
    maximum = "2147483647"
  )
  hakemusKoskee: Int,
  @(Schema @field)(
    example = "1000",
    requiredMode = RequiredMode.REQUIRED,
    minimum = "-9223372036854775807",
    maximum = "9223372036854775807"
  )
  formId: Long,
  @(Schema @field)(
    example = "de4ffbea-1763-4a43-a24d-50ee48b81ff1",
    requiredMode = RequiredMode.NOT_REQUIRED,
    maxLength = 36
  )
  esittelijaId: Option[UUID],
  @(Schema @field)(
    example = "1.2.246.562.24.00000000000000006666",
    requiredMode = RequiredMode.NOT_REQUIRED,
    maxLength = 40
  )
  esittelijaOid: Option[UserOid],
  @(Schema @field)(
    example = "de4ffbea-1763-4a43-a24d-50ee48b81ff1",
    requiredMode = RequiredMode.NOT_REQUIRED,
    maxLength = 36
  )
  asiakirjaId: Option[UUID],
  @(Schema @field)(
    example = "OPH-197-2025",
    requiredMode = RequiredMode.NOT_REQUIRED,
    maxLength = 40
  )
  asiatunnus: Option[String],
  @(Schema @field)(
    example = "AlkukasittelyKesken",
    requiredMode = RequiredMode.REQUIRED,
    maxLength = 50
  )
  kasittelyVaihe: KasittelyVaihe,
  @(Schema @field)(
    example = "2025-06-14T10:59:47.597",
    requiredMode = RequiredMode.NOT_REQUIRED,
    maxLength = 50
  )
  muokattu: Option[LocalDateTime],
  @(Schema @field)(
    example = "false",
    requiredMode = RequiredMode.NOT_REQUIRED
  )
  yhteistutkinto: Boolean,
  @(Schema @field)(
    example = "OPH-197-2025",
    requiredMode = RequiredMode.NOT_REQUIRED,
    maxLength = 40
  )
  lopullinenPaatosVastaavaEhdollinenAsiatunnus: Option[String],
  @(Schema @field)(
    example = "maatjavaltiot2_752",
    requiredMode = RequiredMode.NOT_REQUIRED,
    maxLength = 255
  )
  lopullinenPaatosVastaavaEhdollinenSuoritusmaaKoodiUri: Option[String],
  @(Schema @field)(
    example = "Esittelijän huomioita",
    requiredMode = RequiredMode.NOT_REQUIRED
  )
  esittelijanHuomioita: Option[String]
)

trait UpdatedFromAtaru {
  def hakemusOid: String
  def hakemusKoskee: Int
  def kasittelyVaihe: KasittelyVaihe
}

case class Hakemus(
  hakemusOid: String,
  lomakeOid: String,
  lomakeId: Long,
  lomakkeenKieli: String,
  hakija: Hakija,
  sisalto: Seq[SisaltoItem],
  liitteidenTilat: Seq[AttachmentReview],
  hakemusKoskee: Int,
  asiatunnus: Option[String] = None,
  kirjausPvm: Option[LocalDateTime] = None,
  esittelyPvm: Option[LocalDateTime] = None,
  paatosPvm: Option[LocalDateTime] = None,
  esittelijaOid: Option[String] = None,
  ataruHakemuksenTila: AtaruHakemuksenTila,
  kasittelyVaihe: KasittelyVaihe,
  muokattu: Option[LocalDateTime] = None,
  muutosHistoria: Seq[MuutosHistoriaItem] = Seq.empty,
  taydennyspyyntoLahetetty: Option[LocalDateTime] = None,
  yhteistutkinto: Boolean = false,
  asiakirja: Option[Asiakirja] = None,
  lopullinenPaatosVastaavaEhdollinenAsiatunnus: Option[String] = None,
  lopullinenPaatosVastaavaEhdollinenSuoritusmaaKoodiUri: Option[String] = None,
  esittelijanHuomioita: Option[String]
) extends UpdatedFromAtaru

/**
 * Täysi hakemuksen päivityspyyntö.
 * Käytetään PUT-endpointissa kaikkien käyttäjän muokattavien kenttien korvaamiseen.
 * NULL-arvot pyynnössä → NULL tietokantaan (ei tarvita field wrappereita).
 */
case class HakemusUpdateRequest(
  hakemusKoskee: Int,
  asiatunnus: Option[String],
  kirjausPvm: Option[LocalDateTime],
  esittelyPvm: Option[LocalDateTime],
  paatosPvm: Option[LocalDateTime],
  esittelijaOid: Option[String],
  kasittelyVaihe: KasittelyVaihe,
  yhteistutkinto: Boolean,
  asiakirja: Asiakirja,
  lopullinenPaatosVastaavaEhdollinenAsiatunnus: Option[String],
  lopullinenPaatosVastaavaEhdollinenSuoritusmaaKoodiUri: Option[String],
  esittelijanHuomioita: Option[String]
)

case class AsiatunnusUpdateRequest(
  asiatunnus: String
)

File: tutu-frontend/src/lib/types/hakemus.ts
import {
  Language,
  TranslatedName,
} from '@/src/lib/localization/localizationTypes';
import { Hakija } from '@/src/lib/types/hakija';

export type MuutosHistoriaItem = {
  role: 'Esittelija' | 'Hakija' | 'Irrelevant';
  time: string;
  modifiedBy: string;
};

export enum HakemusKoskee {
  TUTKINNON_TASON_RINNASTAMINEN = 0,
  KELPOISUUS_AMMATTIIN = 1,
  TUTKINTO_SUORITUS_RINNASTAMINEN = 2,
  RIITTAVAT_OPINNOT = 3,
  KELPOISUUS_AMMATTIIN_AP = 4,
  LOPULLINEN_PAATOS = 5,
}

export type Hakemus = {
  hakemusOid: string;
  lomakeOid: string;
  lomakeId: number;
  lomakkeenKieli: Language;
  hakemusKoskee: HakemusKoskee;
  readonly hakija: Hakija;
  asiatunnus: string;
  kirjausPvm: string;
  esittelyPvm: string;
  paatosPvm: string;
  esittelijaOid: string;
  ataruHakemuksenTila: string;
  kasittelyVaihe: string;
  muokattu: string;
  muokkaaja: string;
  muutosHistoria: MuutosHistoriaItem[];
  taydennyspyyntoLahetetty?: string;
  ataruHakemustaMuokattu?: string;
  sisalto: SisaltoItem[];
  liitteidenTilat: TarkistuksenTila[];
  asiakirja: AsiakirjaTieto;
  yhteistutkinto: boolean;
  lopullinenPaatosVastaavaEhdollinenAsiatunnus?: string;
  lopullinenPaatosVastaavaEhdollinenSuoritusmaaKoodiUri?: string;
  esittelijanHuomioita?: string;
};

/**
 * Täysi hakemuksen päivityspyyntö.
 * Käytetään PUT-endpointissa kaikkien käyttäjän muokattavien kenttien korvaamiseen.
 * NULL-arvot pyynnössä → NULL tietokantaan.
 */
export type HakemusUpdateRequest = {
  hakemusKoskee: HakemusKoskee;
  asiatunnus: string | null;
  kirjausPvm: string | null;
  esittelyPvm: string | null;
  paatosPvm: string | null;
  esittelijaOid: string | null;
  kasittelyVaihe: string;
  yhteistutkinto: boolean;
  asiakirja: AsiakirjaTieto;
  lopullinenPaatosVastaavaEhdollinenAsiatunnus: string | null;
  lopullinenPaatosVastaavaEhdollinenSuoritusmaaKoodiUri: string | null;
  esittelijanHuomioita: string | null;
};

export type AsiakirjaTietoUpdateCallback = (
  patch: Partial<AsiakirjaTieto>,
  showUpdateIndicator?: boolean,
) => void;

export type AsiakirjaTieto = {
  pyydettavatAsiakirjat: AsiakirjaPyynto[];
  allekirjoituksetTarkistettu: boolean;
  allekirjoituksetTarkistettuLisatiedot: string | null | undefined;
  alkuperaisetAsiakirjatSaatuNahtavaksi: boolean;
  alkuperaisetAsiakirjatSaatuNahtavaksiLisatiedot: string | null | undefined;
  selvityksetSaatu: boolean;
  viimeinenAsiakirjaHakijalta?: string;
  asiakirjamallitTutkinnoista?: AsiakirjamallitTutkinnoista;
  imiPyynto: ImiPyynto;
  apHakemus?: boolean | null;
  suostumusVahvistamiselleSaatu: boolean;
  valmistumisenVahvistus: ValmistumisenVahvistus;
  huomiotMuistioon?: string;
  esittelijanHuomioita?: string;
};

export type InfoText = {
  label?: TranslatedName;
  value?: TranslatedName;
};

export type SisaltoPathNode = SisaltoItem | SisaltoValue;

export type SisaltoItem = {
  key: string;
  fieldType: string;
  value: SisaltoValue[];
  label: TranslatedName;
  children: SisaltoItem[];
  params?: SisaltoItemInfoText;
  previous?: SisaltoPathNode;
  infoText?: InfoText;
};

export type SisaltoValue = {
  label: TranslatedName;
  value: string;
  followups: SisaltoItem[];
  formId?: string;
  previous?: SisaltoPathNode;
};

export type SisaltoItemInfoText = {
  label?: TranslatedName;
  value?: TranslatedName;
};

export type AsiakirjaMetadata = {
  key: string;
  filename: string;
  saapumisaika?: string;
};

export type TarkistuksenTila = {
  attachment: string;
  state: string;
  hakukohde: string;
  updateTime?: string;
};

export type AsiakirjaPyynto = {
  id: string;
  asiakirjanTyyppi: string;
};

export type AsiakirjamalliLahde =
  | 'ece'
  | 'UK_enic'
  | 'naric_portal'
  | 'nuffic'
  | 'aacrao'
  | 'muu';

export type AsiakirjamalliTutkinnosta = {
  lahde: AsiakirjamalliLahde;
  vastaavuus: boolean;
  kuvaus?: string;
};

export type AsiakirjamallitTutkinnoista = Partial<
  Record<AsiakirjamalliLahde, AsiakirjamalliTutkinnosta>
>;

export type ImiPyynto = {
  imiPyynto?: boolean | null;
  imiPyyntoNumero?: string | null;
  imiPyyntoLahetetty?: string | null;
  imiPyyntoVastattu?: string | null;
};

export type ValmistumisenVahvistusVastaus =
  | 'Myonteinen'
  | 'Kielteinen'
  | 'EiVastausta';

export type ValmistumisenVahvistus = {
  valmistumisenVahvistus: boolean;
  valmistumisenVahvistusPyyntoLahetetty: string | null;
  valmistumisenVahvistusSaatu: string | null;
  valmistumisenVahvistusVastaus: ValmistumisenVahvistusVastaus | null;
  valmistumisenVahvistusLisatieto: string | null;
};
